module/VuFind/src/VuFind/Db/Service/RatingsService.php

```diff
@@ -40,6 +42,156 @@
+            $parameters['user'] = $this->userService->getUserById($userId);
+        }
+        $dql .= ' WHERE ' . implode(' AND ', $dqlWhere);
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result = $query->getResult();
+        return [
+            'count' => $result[0]['count'],
+            'rating' => floor($result[0]['rating']) ?? 0
+        ];
+    }
+
+    /**
+     * Get rating breakdown for the specified resource.
+     *
+     * @param string $id     Record ID to look up
+     * @param string $source Source of record to look up
+     * @param array  $groups Group definition (key => [min, max])
+     *
+     * @return array Array with keys count and rating (between 0 and 100) as well as
+     * an groups array with ratings from lowest to highest
+     */
+    public function getCountsForResource(
+        string $id,
+        string $source,
+        array $groups
+    ): array {
+        $result = [
+            'count' => 0,
+            'rating' => 0,
+            'groups' => [],
+        ];
+        foreach (array_keys($groups) as $key) {
+            $result['groups'][$key] = 0;
+        }
+
+        $resource = $this->resourceService->findResource($id, $source);
+        if (empty($resource)) {
+            return $result;
+        }
+        $dql = "SELECT COUNT(r.id) AS count, r.rating AS rating "
+            . "FROM " . $this->getEntityClass(Ratings::class) . " r "
+            . "WHERE r.resource = :resource "
+            . "GROUP BY rating";
+
+        $parameters['resource'] = $resource;
+
+        $query = $this->entityManager->createQuery($dql);
+
+        $query->setParameters($parameters);
+        $queryResult = $query->getResult();
+
+        $ratingTotal = 0;
+        $groupCount = 0;
+        foreach ($queryResult as $rating) {
+            $result['count'] += $rating['count'];
+            $ratingTotal += $rating['rating'];
+            ++$groupCount;
+            if ($groups) {
+                foreach ($groups as $key => $range) {
+                    if ($rating['rating'] >= $range[0]
+                        && $rating['rating'] <= $range[1]
+                    ) {
+                        $result['groups'][$key] = ($result['groups'][$key] ?? 0)
+                            + $rating['count'];
+                    }
+                }
+            }
+        }
+        $result['rating'] = $groupCount ? floor($ratingTotal / $groupCount) : 0;
+        return $result;
+    }
+
+    /**
+     * Deletes all ratings by a user.
+     *
+     * @param \VuFind\Db\Row\User $user User object
+     *
+     * @return void
+     */
+    public function deleteByUser(\VuFind\Db\Row\User $user): void
+    {
+        $dql = 'DELETE FROM ' . $this->getEntityClass(Ratings::class) . ' r '
+            . "WHERE r.user = :user";
+        $userEntity = $this->userService->getUserById($user->id);
+        $parameters['user'] = $userEntity;
```

Comment:
Can you just use $user->id here without having to fetch the user entity? I think usually you can pass an ID value instead of a full entity in this situation. Maybe we can eliminate the dependency on the user service.